### DISCUSSION TITLE
zod OR doesn't work as expected when parsing.

### DISCUSSION THREAD

**Discussion by jarrodmedrano** [colinhacks/zod#2954] [Q&A] [answered]
**Link to Repl**: https://replit.com/@jarrodmedrano/zod-Or-bug-Optional#index.ts

I have two types, animal and person. Animal species is optional.

```ts
const personResult = z.object({
  name: z.string()
});

// when species optional, result gives nothing {}
const animalResult = z.object({
  species: z.string().optional(),
});

// when species is not optional, result will be correct
const animalResultWorks = z.object({
  species: z.string(),
});

```
When you parse them...
```ts
const aPerson = {
  name: "Jerry"
}

const animalorPerson = z.object({
  something: animalResult.or(personResult).optional(),
}).optional();

const parsed = animalorPerson.safeParse({
  something: aPerson
}); 
```
You will get this result:
```ts
 { success: true, data: { something: {} } }
 ```

What it should be:
```ts
{
  success: true,
  data: { something: { name: 'Jerry' } }
}
```

You will get the correct result if you take the 'optional' off of species.
You will get the correct result if you reverse the order of the .or operation
if Species is optional and aPerson contains something totally invalid, you won't see any error message.

**Comment by JacobWeisenburger** [comment, marked as answer]
https://github.com/colinhacks/zod#unions

> Zod will test the input against each of the "options" in order and return the first value that validates successfully.

This seems to be working as expected.

---

try using `.strict()`
https://github.com/colinhacks/zod#strict

```ts
const personResult = z.object( {
    name: z.string()
} )

const animalResult = z.object( {
    species: z.string().optional(),
} ).strict()

const animalOrPerson = z.object( {
    something: animalResult.or( personResult ).optional(),
} ).optional()

console.log(
    animalOrPerson.parse( {
        something: {
            name: 'Jerry'
        }
    } )
)
// {
//     something: {
//       name: "Jerry"
//     }
// }
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger
